src/cli/Cli.ts

```diff
@@ -158,110 +158,109 @@ export class Cli {
-            listener.stderr(message);
-          }
-          logErr.push(message);
+    const log: string[] = [];
+    const logErr: string[] = [];
+    const logger: Logger = {
+      log: (message: any): void => {
+        const formattedMessage = Cli.formatOutput(message, args.options);
+        if (listener && listener.stdout) {
+          listener.stdout(formattedMessage);
         }
-      };
-
-      if (args.options.debug) {
-        const message = `Executing command ${command.name} with options ${JSON.stringify(args)}`;
+        log.push(formattedMessage);
+      },
+      logRaw: (message: any): void => {
+        const formattedMessage = Cli.formatOutput(message, args.options);
+        if (listener && listener.stdout) {
+          listener.stdout(formattedMessage);
+        }
+        log.push(formattedMessage);
+      },
+      logToStderr: (message: any): void => {
         if (listener && listener.stderr) {
           listener.stderr(message);
         }
         logErr.push(message);
       }
+    };
 
-      // store the current command name, if any and set the name to the name of
-      // the command to execute
-      const cli = Cli.getInstance();
-      const parentCommandName: string | undefined = cli.currentCommandName;
-      cli.currentCommandName = command.getCommandName();
-      // store the current logger if any
-      const currentLogger: Logger | undefined = request.logger;
-
-      command.action(logger, args as any, (err: any): void => {
-        // restore the original command name
-        cli.currentCommandName = parentCommandName;
-        // restore the original logger
-        request.logger = currentLogger;
-
-        if (err) {
-          return reject({
-            error: err,
-            stderr: logErr.join(os.EOL)
-          });
-        }
+    if (args.options.debug) {
+      const message = `Executing command ${command.name} with options ${JSON.stringify(args)}`;
+      if (listener && listener.stderr) {
+        listener.stderr(message);
+      }
+      logErr.push(message);
+    }
 
-        resolve({
-          stdout: log.join(os.EOL),
-          stderr: logErr.join(os.EOL)
-        });
+    // store the current command name, if any and set the name to the name of
+    // the command to execute
+    const cli = Cli.getInstance();
+    const parentCommandName: string | undefined = cli.currentCommandName;
+    cli.currentCommandName = command.getCommandName();
+    // store the current logger if any
+    const currentLogger: Logger | undefined = request.logger;
+
+    try {
+      await command.action(logger, args as any);
+
+      // restore the original command name
+      cli.currentCommandName = parentCommandName;
+      // restore the original logger
+      request.logger = currentLogger;
+
+      return ({
+        stdout: log.join(os.EOL),
+        stderr: logErr.join(os.EOL)
       });
-    });
+    }
+    catch (err: any) {
+      // restore the original command name
```

Comment:
Is there a way for us to move restoring the original name and logger to a `finally` clause or is this where you hit the coverage issue and you had to work around it with the current approach? If that's indeed the case, let's add a comment here so that we know why it's done that way and can hopefully fix it in the future.

Reply:
I used to have a `finally` there indeed. But because an error is thrown in the `catch` branch, Istanbul thinks that the `finally` is never hit. If you don't throw an exception in the `catch` branch, Istanbul works as expected.
This is indeed the reason why I copied the code in the `try` and `catch` branch.
Adding a comment seems like a good idea :+1:

Reply:
Something like this?

```
// restoring the command and logger is done here instead of in a 'finally' because there were issues with the code coverage tool
```